components/campaign/CampaignChaosBagView/index.tsx

```diff
@@ -0,0 +1,382 @@
+    const list = keys(chaosBag);
+    const weight = values(chaosBag);
+
+    const totalWeight = weight.reduce((prev, cur) => {
+      return (prev && cur) ? (prev + cur) : 0;
+    });
+
+    if (totalWeight) {
+      const random_num = this._getRandomNumber(0, totalWeight);
+      let weight_sum = 0;
+      const length = list.length;
+      for (let i = 0; i < length; i++) {
+        const item = weight[i];
+        if (item) {
+          weight_sum += item;
+          weight_sum = +weight_sum.toFixed(2);
+        }
+
+        if (random_num <= weight_sum) {
+          const tokenName = list[i];
+          let drawnToken = null;
+          CHAOS_TOKENS.forEach(token => {
+            if (token === tokenName) {
+              drawnToken = token;
+            }
+          });
+
+          if (drawnToken) {
+            return drawnToken;
+          }
+        }
+      }
+    }
+  }
+
+  _drawToken = (refresh = false) => {
+    const { chaosBagResults } = this.state;
+    const { chaosBag } = this.props;
+
+    if (!chaosBagResults.currentToken || refresh) {
+      const newIconKey = this._getRandomChaosToken(chaosBag);
+      this.setState({
+        chaosBagResults: {
+          currentToken: newIconKey,
+          drawnTokens: chaosBagResults.drawnTokens,
+          total: chaosBagResults.total + 1,
+        },
+      });
+    } else {
+      this.setState({
+        chaosBagResults: {
+          currentToken: undefined,
+          drawnTokens: [],
+          total: chaosBagResults.total,
+        },
+      });
+    }
+  }
+
+  _renderChaosToken = (iconKey?: ChaosTokenType) => {
+    return (
+      <TouchableOpacity onPress={this._handleDrawTokenPressed}>
+        <ChaosToken iconKey={iconKey} />
+      </TouchableOpacity>
+    );
+  }
+
+  _addTokenAndDraw = () => {
+    const {
+      chaosBagResults,
+    } = this.state;
+
+    const drawnTokens = chaosBagResults.drawnTokens;
+    if (chaosBagResults.currentToken) {
+      drawnTokens.push(chaosBagResults.currentToken);
+    }
+
+    this.setState({
+      chaosBagResults: {
+        currentToken: chaosBagResults.currentToken,
+        drawnTokens: drawnTokens,
+        total: chaosBagResults.total,
+      },
+    });
+    this._drawToken(true);
+  }
+
+  _clearTokens = () => {
+    const { chaosBagResults } = this.state;
+
+    this.setState({
+      chaosBagResults: {
+        currentToken: undefined,
+        drawnTokens: [],
+        total: chaosBagResults.total,
+      },
+    });
+  }
+
+  _renderDrawnTokens() {
+    const { chaosBagResults } = this.state;
+
+    const drawnTokens = chaosBagResults.drawnTokens;
+
+    if (drawnTokens.length > 0) {
+      return drawnTokens.map(function(token, index) {
+        return (
+          <ChaosToken key={index} iconKey={token} small />
+        );
+      });
+    }
+
+    return <Text style={[styles.drawTokenText, typography.text]}>Tap token to draw</Text>;
+  }
+
+  _renderDrawButton = () => {
```

Comment:
Since we don't usually pass the render functions, I just keep them normal most often.

```
renderDrawButton() {
}
```